CB-10332 Android Guide Improvements

riknoll opened the pull request:
Lot's of changes, including:

* Fixing Markdown, links, grammar, etc.
* Removing outdated and duplicated content
* Removing content where we should just point to the Android docs (like setting up an emulator)

Things left to do:

* Combine all the pages into one page
* Completely rewrite the guide for embedding a webview (it is 100% outdated)
* Remove pages that shouldn't exist anymore (Android's config.xml got integrated into the main doc)

vladimir-kotikov commented on www/docs/en/dev/guide/platforms/android/index.md:
```diff
@@ -32,276 +32,152 @@ development. For a comparison of the two development paths, see the
+Install [Java Development Kit (JDK) 7](http://www.oracle.com/technetwork/java/javase/downloads/jdk7-downloads-1880260.html)
+or later.
+
+When installing on Windows you also need to set `JAVA_HOME` Environment Variable
```
I'm not 100% sure, but this is someting that should be set by `check_reqs` at the [build time](https://github.com/apache/cordova-android/blob/master/bin/lib/check_reqs.js#L99)

vladimir-kotikov commented on www/docs/en/dev/guide/platforms/android/index.md:
```diff
@@ -32,276 +32,152 @@ development. For a comparison of the two development paths, see the
+#### Update your PATH
+
+For Cordova's CLI and command line tools to work correctly, you will need to
+include the SDK's `tools` and `platform-tools` directories in your `PATH`.
```
The same as for `JAVA_HOME` - this should be [set by `check_reqs` automatically](https://github.com/apache/cordova-android/blob/master/bin/lib/check_reqs.js#L166)

riknoll commented:
@vladimir-kotikov Thanks for the heads up, I wasn't aware of that! After discussing this with @nikhilkh, I think I'm going to leave the instructions in but add that Cordova will try to do it automatically. Seems like that script might not always work.

vladimir-kotikov commented:
@riknoll, +1, makes sense

rakatyal commented on www/docs/en/dev/guide/platforms/android/index.md:
```diff
@@ -32,276 +32,152 @@ development. For a comparison of the two development paths, see the
-Install the [Android Stand-alone SDK Tools](http://developer.android.com/sdk/installing/index.html?pkg=tools) or [Android Studio](http://developer.android.com/sdk/installing/index.html?pkg=studio). Procceed with `Android Studio` if you plan 
+Install the [Android Stand-alone SDK Tools](http://developer.android.com/sdk/installing/index.html?pkg=tools) or [Android Studio](http://developer.android.com/sdk/installing/index.html?pkg=studio).
+Proceed with Android Studio if you plan on
 developing new Cordova for Android plugins or using native tools to
```
new Cordova apps or Android plugins?

riknoll replied:
For some reason this article says "Cordova for Android" all over the place and so I kept that up. This sentence was referring to plugins. Android Studio can be used with apps, but  I think the better use case is with plugins.

rakatyal replied:
Yeah all right.

rakatyal commented on www/docs/en/dev/guide/platforms/android/index.md:
```diff
@@ -32,276 +32,152 @@ development. For a comparison of the two development paths, see the
+  1. For the `Gradle Sync` question you can simply answer __Yes__.
 
-You are all set now and can build and run the app directly from `Android Studio`.
+Once it finishes importing, you should be able build and run the app directly
```
able to

rakatyal commented on www/docs/en/dev/guide/platforms/android/index.md:
```diff
@@ -32,276 +32,152 @@ development. For a comparison of the two development paths, see the
+the command:
 
-The following generates a detailed log of the app as it runs:
+    $ cordova run --list
```
Can we list the expected output?

riknoll replied:
It's pretty self explanatory. Do you think we need to?
```
Available android devices:
Available android virtual devices:
<emulator_name>
```

riknoll replied:
Doesn't hurt, I guess.

rakatyal replied:
Yeah, on second thought, let's skip it.

rakatyal commented on www/docs/en/dev/guide/platforms/android/plugin.md:
```diff
@@ -31,12 +31,12 @@ platform and back.  For another sample, see also the comments in
-should also handle `[pause](../../../cordova/events/events.pause.html)` and `[resume](../../../cordova/events/events.resume.html)` events, along with any message
-passing between plugins.  Plugins with long-running requests,
+overriding one of its `execute` methods. As a best practice, the plugin
+should also handle [`pause`](../../../cordova/events/events.pause.html) and [`resume`](../../../cordova/events/events.resume.html) events, along with any message
```
Link these to events reference?

nikhilkh replied:
Should you link to android lifecyle here?

riknoll replied:
@rakatyal Yep, I'll do that
@nikhilkh Good point!

rakatyal commented on www/docs/en/dev/guide/platforms/android/plugin.md:
```diff
@@ -118,7 +118,8 @@ exception names as much as possible.
 `WebView` interface; instead, it runs on the `WebCore` thread, as
 does the `execute` method.  If you need to interact with the user
-interface, you should use the following variation:
+interface, you should use the [Activity's `runOnUiThread`][ref-runonuithread]
```
This appears very weird on the github rich diff. I hope it's better on the website?

rakatyal commented on www/docs/en/dev/guide/platforms/android/tools.md:
```diff
@@ -19,206 +19,218 @@ license: >
-Run the `create` command, specifying the existing path to the project,
-the reverse-domain-style package identifier, and the app's display
-name.  Here is the syntax for both Mac/Linux and Windows:
+  * **cdvBuildMultipleApks** (default: false)
```
Can we have these in a table? Makes it more readable.

nikhilkh commented on www/docs/en/dev/guide/platforms/android/plugin.md:
```diff
@@ -72,7 +72,7 @@ Otherwise, the plugin may compile but still be unavailable to Cordova.
+`name` attribute is set to `"true"` in `config.xml`. For example,
 
     <feature name="Echo">
         <param name="android-package" value="<full_name_including_namespace>" />
```
What does the `android-package` stand for?

riknoll replied:
I believe the `"android-package"` param is used for defining the Java package for your app so that the directory structure gets created correctly. See an example usage here: https://github.com/apache/cordova-plugin-camera/blob/master/plugin.xml#L64

nikhilkh commented on www/docs/en/dev/guide/platforms/android/index.md:
```diff
@@ -32,276 +32,152 @@ development. For a comparison of the two development paths, see the
+
+cordova-android Version | Supported Android API-Levels
+------------------------|-----------------------------
+5.X.X                   | 14 - 23
```
I don't have better ideas but this will certainly require some maintenance every time we change the API level.

nikhilkh commented on www/docs/en/dev/guide/platforms/android/plugin.md:
```diff
@@ -155,27 +158,24 @@ thread, but do not want to block the `WebCore` thread either:
+This allows commonly used libraries such as _gson_, _android-support-v4_,
+and _google-play-services_ to be used by multiple plugins without conflict.
+
+The second option is to use the `<lib-file />` tag to specify the location of
```
Does `framework` support a jar reference too? Are these mutually exclusive or `framework` is always preferred over `lib-file`?

nikhilkh commented:
Should we add a section on "Diagnosing failures" - it could be part of a subsequent PR - in particular, talking about adb logcat or referencing appropriate Android docs.

riknoll commented:
@nikhilkh yes, I think the plan is to add a debugging section.

riknoll commented:
Okay, I just forced pushed some major changes to this PR:

* Addressed most of the code review
* Deleted old pages and combined other pages into single page guide
* Lots of miscellaneous content fixes and reorganization to fit the single page platform guide structure
* Rebased to master

Plugin guide is still separate, since that is going to move. Lifecycle is also separate, because I need to figure out what to do about that since I've been handing out links to that page all over the place because of the save-restore API changes.

riknoll commented:
@rakatyal @nikhilkh @vladimir-kotikov  @infil00p please look this over. I have some more updates to content that are coming, like [CB-10206](https://issues.apache.org/jira/browse/CB-10206). I just want to get the review started now.

riknoll commented:
I moved the lifecycle guide into the single page guide and added redirects to our `.htaccess` file that handle requests for all of the pages that got combined.

omefire commented on www/docs/en/dev/guide/platforms/android/index.md:
```diff
@@ -30,278 +30,676 @@ platform-centered shell tools or cross-platform Cordova CLI for
 development. For a comparison of the two development paths, see the
 [Overview](../../overview/index.html).  For details on the CLI, see [The Command-Line Interface](../../cli/index.html).
```
Link should point to: http://localhost:3000/docs/en/dev/guide/overview/index.html#development-paths

omefire commented on www/docs/en/dev/guide/platforms/android/index.md:
```diff
@@ -30,278 +30,676 @@ platform-centered shell tools or cross-platform Cordova CLI for
 
-Cordova for Android requires the Android SDK which could be installed
+Cordova for Android requires the Android SDK which can be installed
 on OS X, Linux or Windows operation system. See the Android SDK's
```
'operation' => 'operating'

riknoll commented:
Rebased to master and updated to fix the redirects merged in with #494

omefire commented on www/docs/en/dev/guide/platforms/android/plugin.md:
```diff
@@ -23,24 +23,17 @@ title: Android Plugins
 
 This section provides details for how to implement native plugin code
-on the Android platform. Before reading this, see Application Plugins
+on the Android platform. Before reading this, see the [Plugin Development Guide][plugin-dev]
```
This section should probably be renamed: 'Android Plugin Development Guide'.

riknoll replied:
Is that our official naming convention? I'm inclined to leave it as is until we combine all the plugin guides into one.

omefire replied:
I think we should rename it. any ideas? @rakatyal

rakatyal replied:
I am okay with anything as long as it's consistent. 'Android Development Guide' and 'Android Plugin Development Guide' and so on for each platform?

riknoll replied:
Done

rakatyal commented:
Are we adding link to the 'Upgrading' article?

riknoll commented:
@rakatyal I thought we were deleting the upgrading article

rakatyal commented:
We are not deleting it but referencing it from the main guide.

riknoll commented:
Alright, I'll resurrect it and add some stuff about 5.0.0

rakatyal commented:
:ship: it!
